File: VertexTool.m
/*
Copyright (C) 2010-2012 Kristian Duske

This file is part of TrenchBroom.

TrenchBroom is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

TrenchBroom is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with TrenchBroom.  If not, see <http://www.gnu.org/licenses/>.
*/

#import "VertexTool.h"
#import "Math.h"
#import "MapWindowController.h"
#import "MapDocument.h"
#import "CursorManager.h"
#import "DragVertexCursor.h"
#import "PickingHit.h"
#import "PickingHitList.h"
#import "Camera.h"
#import "EditingSystem.h"
#import "Options.h"
#import "Grid.h"
#import "SelectionManager.h"
#import "MutableBrush.h"
#import "Face.h"

@interface VertexTool (private)

- (BOOL)isAlternatePlaneModifierPressed;
- (void)updateMoveDirectionWithRay:(const TRay *)theRay hits:(PickingHitList *)theHits;

@end

@implementation VertexTool (private)

- (BOOL)isAlternatePlaneModifierPressed {
    return keyStatus == KS_OPTION;
}

- (void)updateMoveDirectionWithRay:(const TRay *)theRay hits:(PickingHitList *)theHits {
    if (editingSystem != nil)
        [editingSystem release];

    Camera* camera = [windowController camera];
    editingSystem = [[EditingSystem alloc] initWithCamera:camera vertical:[self isAlternatePlaneModifierPressed]];
}

@end

@implementation VertexTool

- (id)initWithWindowController:(MapWindowController *)theWindowController {
    NSAssert(theWindowController != nil, @"window controller must not be nil");
    
    if ((self = [self init])) {
        windowController = theWindowController;
        cursor = [[DragVertexCursor alloc] init];
        state = VTS_DEFAULT;
    }
    
    return self;
}

- (void)dealloc {
    [cursor release];
    [editingSystem release];
    [super dealloc];
}

- (void)handleKeyStatusChanged:(NSEvent *)event status:(EKeyStatus)theKeyStatus ray:(TRay *)ray hits:(PickingHitList *)hits {
    keyStatus = theKeyStatus;
}

- (void)beginLeftDrag:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    NSAssert(event != NULL, @"event must not be NULL");
    NSAssert(ray != NULL, @"ray must not be NULL");
    NSAssert(hits != nil, @"hit list must not be nil");
    
    [hits retain];
    
    PickingHit* hit = [hits firstHitOfType:HT_VERTEX ignoreOccluders:NO];
    if (hit != nil) {
        MapDocument* map = [windowController document];
        NSUndoManager* undoManager = [map undoManager];
        [undoManager setGroupsByEvent:NO];
        [undoManager beginUndoGrouping];
        
        brush = [hit object];
        [map snapBrushes:[NSArray arrayWithObject:brush]];
        
        index = [hit vertexIndex];
        const TVertexList* vertices = [brush vertices];
        const TEdgeList* edges = [brush edges];
        
        if (index < vertices->count) {
            TVertex* vertex = vertices->items[index];
            lastPoint = vertex->position;
        } else if (index < vertices->count + edges->count) {
            TEdge* edge = edges->items[index - vertices->count];
            centerOfEdge(edge, &lastPoint);
        } else {
            // the side index is not necessarily the same as the face index!!!
            id <Face> face = [[brush faces] objectAtIndex:index - edges->count - vertices->count];
            centerOfVertices([face vertices], &lastPoint);
        }
        editingPoint = lastPoint;
        
        state = VTS_DRAG;
    }
    
    [hits release];
}

- (void)leftDrag:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    if (state != VTS_DRAG)
        return;
    
    float dist = [editingSystem intersectWithRay:ray planePosition:&editingPoint];
    if (isnan(dist))
        return;
    
    TVector3f point;
    rayPointAtDistance(ray, dist, &point);
    
    Options* options = [windowController options];
    Grid* grid = [options grid];
    [grid snapToGridV3f:&point result:&point];

    TVector3f deltaf;
    subV3f(&point, &lastPoint, &deltaf);

    if (nullV3f(&deltaf))
        return;
    
    MapDocument* map = [windowController document];
    index = [map dragVertex:index brush:brush delta:&deltaf];
    if (index == -1) {
        [self endLeftDrag:event ray:ray hits:hits];
        state = VTS_CANCEL;
        [self unsetCursor:event ray:ray hits:hits];
    } else if (index < [brush vertices]->count) {
        lastPoint = point;
    }
}

- (void)endLeftDrag:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    if (state == VTS_CANCEL) {
        [self setCursor:event ray:ray hits:hits];
        state = VTS_DEFAULT;
    }
    
    if (state != VTS_DRAG)
        return;
    
    MapDocument* map = [windowController document];
    NSUndoManager* undoManager = [map undoManager];
    [undoManager setActionName:[self actionName]];
    [undoManager endUndoGrouping];
    [undoManager setGroupsByEvent:YES];

    brush = nil;
    index = -1;
    
    state = VTS_DEFAULT;
}

- (void)setCursor:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    CursorManager* cursorManager = [windowController cursorManager];
    [cursorManager pushCursor:cursor];
    [self updateCursor:event ray:ray hits:hits];
}

- (void)unsetCursor:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    CursorManager* cursorManager = [windowController cursorManager];
    [cursorManager popCursor];
}

- (void)updateCursor:(NSEvent *)event ray:(TRay *)ray hits:(PickingHitList *)hits {
    TVector3f position;

    [self updateMoveDirectionWithRay:ray hits:hits];
    
    if (state == VTS_DRAG) {
        float dist = [editingSystem intersectWithRay:ray planePosition:&editingPoint];
        if (isnan(dist))
            return;
        
        rayPointAtDistance(ray, dist, &position);

        Camera* camera = [windowController camera];
        [cursor setEditingSystem:editingSystem];
        [cursor setCameraPosition:[camera position]];
        [cursor setAttention:NO];
    } else if (state == VTS_CANCEL) {
        
    } else {
        NSArray* vertexHits = [hits hitsOfType:HT_VERTEX];
        if ([vertexHits count] == 0)
            return;
        
        position = *[[vertexHits objectAtIndex:0] hitPoint];
        
        BOOL attention = NO;
        NSEnumerator* hitEn = [vertexHits objectEnumerator];
        PickingHit* hit;
        while (!attention && (hit = [hitEn nextObject])) {
            id <Brush> hitBrush = [hit object];
            const TVertexList* vertices = [hitBrush vertices];
            for (int i = 0; i < vertices->count && !attention; i++)
                attention = !intV3f(&vertices->items[i]->position);
        }
        
        Camera* camera = [windowController camera];
        [cursor setEditingSystem:editingSystem];
        [cursor setCameraPosition:[camera position]];
        [cursor setAttention:attention];
    }
    
    [cursor setPosition:&position];
}

- (NSString *)actionName {
    return @"Drag Vertices";
}
@end

File: Core/Source/Controller/VertexTool.cpp
/*
 Copyright (C) 2010-2012 Kristian Duske
 
 This file is part of TrenchBroom.
 
 TrenchBroom is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 TrenchBroom is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with TrenchBroom.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "VertexTool.h"

#include "Controller/Editor.h"
#include "Controller/Grid.h"
#include "Model/Undo/UndoManager.h"
#include "Renderer/Figures/VertexToolFigure.h"

#include <cassert>

namespace TrenchBroom {
    namespace Controller {
        size_t VertexTool::index(Model::Hit& hit) {
            return hit.index;
        }

        void VertexTool::brushesDidChange(const Model::BrushList& brushes) {
            refreshFigure(true);
        }
        
        void VertexTool::mapCleared(Model::Map& map) {
            refreshFigure(true);
        }
        
        void VertexTool::selectionChanged(const Model::SelectionEventData& event) {
            refreshFigure(true);
        }

        VertexTool::VertexTool(Controller::Editor& editor) : DragTool(editor), m_selected(false), m_brush(NULL), m_index(-1), m_figureCreated(false) {
        }

        bool VertexTool::handleActivated(InputEvent& event) {
            assert(state() == TB_TS_DEFAULT);
            
            Model::Map& map = editor().map();
            Model::Selection& selection = map.selection();
            
            map.brushesDidChange        += new BrushListener(this, &VertexTool::brushesDidChange);
            map.mapCleared              += new MapListener(this, &VertexTool::mapCleared);
            selection.selectionAdded    += new SelectionListener(this, &VertexTool::selectionChanged);
            selection.selectionRemoved  += new SelectionListener(this, &VertexTool::selectionChanged);
            
            if (!m_figureCreated) {
                Renderer::VertexToolFigure* figure = new Renderer::VertexToolFigure(*this);
                addFigure(*figure);
                m_figureCreated = true;
            }

            refreshFigure(true);

            return true;
        }
        
        bool VertexTool::handleDeactivated(InputEvent& event) {
            assert(active());

            Model::Map& map = editor().map();
            Model::Selection& selection = map.selection();
            
            map.brushesDidChange        -= new BrushListener(this, &VertexTool::brushesDidChange);
            map.mapCleared              -= new MapListener(this, &VertexTool::mapCleared);
            selection.selectionAdded    -= new SelectionListener(this, &VertexTool::selectionChanged);
            selection.selectionRemoved  -= new SelectionListener(this, &VertexTool::selectionChanged);
            
            refreshFigure(true);
            
            return true;
        }
        
        bool VertexTool::handleMouseDown(InputEvent& event) {
            assert(active());

            if (event.mouseButton != TB_MB_LEFT || !noModifierPressed(event))
                return false;
            
            Model::Hit* hit = event.hits->first(hitType(), true);
            if (hit != NULL) {
                m_brush = &hit->brush();
                m_index = index(*hit);
                m_selected = true;
                refreshFigure(true);
                return true;
            }
            
            return false;
        }
        
        bool VertexTool::handleMouseUp(InputEvent& event) {
            assert(active());

            if (event.mouseButton != TB_MB_LEFT)
                return false;
            
            if (selected()) {
                m_brush = NULL;
                m_index = 0;
                m_selected = false;
                refreshFigure(false);
                return true;
            }
            
            return false;
        }
        
        bool VertexTool::handleBeginPlaneDrag(InputEvent& event, Vec3f& initialPoint) {
            if (event.mouseButton != TB_MB_LEFT || !noModifierPressed(event))
                return false;
            
            Model::Hit* hit = event.hits->first(hitType(), true);
            if (hit == NULL)
                return false;
            
            assert(selected());

            m_brush = &hit->brush();
            m_index = index(*hit);
            initialPoint = hit->hitPoint;
            
            editor().map().undoManager().begin(undoName());
            return true;
        }
        
        bool VertexTool::handlePlaneDrag(InputEvent& event, const Vec3f& lastMousePoint, const Vec3f& curMousePoint, Vec3f& referencePoint) {
            assert(event.mouseButton == TB_MB_LEFT);
            
            if (state() == TB_TS_DRAG) {
                assert(m_brush != NULL);
                
                Vec3f position = movePosition(*m_brush, m_index);
                Vec3f delta = moveDelta(position, curMousePoint - referencePoint);
                
                if (delta.null())
                    return true;
                
                Model::MoveResult result = performMove(*m_brush, m_index, delta);
                m_index = result.index;
                if (result.deleted)
                    return false;
                else if (result.moved)
                    referencePoint += delta;

                refreshFigure(result.moved);
            }
            
            return true;
        }
        
        void VertexTool::handleEndPlaneDrag(InputEvent& event) {
            assert(event.mouseButton == TB_MB_LEFT);
            
            if (state() == TB_TS_DRAG) {
                editor().map().undoManager().end();
                m_brush = NULL;
                m_index = 0;
            }
            
            m_selected = false;
            refreshFigure(true);
        }
    }
}
